### ISSUE TITLE
Problem: http server doesn't support HTTPS

### ISSUE THREAD

**Issue by yrashk** [omnigres/omnigres#405]
This is not because H2O (the HTTP server we use) doesn't (it does); it's because we never added this configuration knob.

However, this is an important step for running the entire stack without additional tools.

Proposed solution: add a way to configure a certificate per listener

> A grand vision here was (and still is) to implement `omni_pki`, a comprehensive extension for managing keys and certificates (generate CSR, store keys, certificates, etc.). It would be essentially a wrapper around some of OpenSSL's functionality. But perhaps this can wait a bit. It is, however, helpful to know the original intention so that at least we know where this can go and be advised by it.

The grand vision notwithstanding, we need to figure out a short path to this configurability. Some options:

1. The `omni_httpd.certificates` (name TBD) table would list actual certificates, and how we match them to listeners can be very interesting: chances are, we can match a certificate by SNI (and maybe listener ID?), making it a super-easy setup. My reading of [h2o](https://github.com/h2o/h2o/blob/master/src/main.c) suggests that's quite possible.
2. Similar to above, but the resolution of the certificate would done through a stored function. This will likely be slower as it involves communication with the database thread. Might not make sense given we want to make it performant.
3. If the grand vision is the one that makes the most sense, let's talk [TODO: write down thoughts on that].

**NB: Before implementing this bounty, please provide a rough plan for review to ensure your effort will be most fruitful.**

Not in scope for this bounty:

* ACME/LetsEncrypt

Other comments:

Eventually, much of omni_httpd functionality will be extracted into omni_workers, but this will still be useful as the HTTP-specific block will remain.

---

**Bounty size: $250**

[Read more about bounties](https://github.com/omnigres/omnigres/wiki/Bounties)

**Comment by bilalshaikh292** [comment]
Heyy @yrashk is the issue still open ? I would like to contribute ?

**Comment by yrashk** [comment]
@bilalshaikh292 it's open!

**Comment by skushagra9** [comment]
Please assign this issue to me ,I would love to contribute

**Comment by UkuLoskit** [comment]
hey, @yrashk, is there a reasonably easy way to get this extension running in a nix environment with the extensions? when I set up the nix environment, it seemed like some of the extensions like httpd were not included and I noticed that there was a TODO regarding this in the nix files. 

The only way I could get a local setup running for now was using Docker by commenting out the Rust part which was pretty slow. Even with these changes the Docker build is quite slow if I want to change some C code (since Docker caching does not work).

**Comment by yrashk** [comment]
> hey, @yrashk, is there a reasonably easy way to get this extension running in a nix environment with the extensions? when I set up the nix environment, it seemed like some of the extensions like httpd were not included and I noticed that there was a TODO regarding this in the nix files. 
> 
> The only way I could get a local setup running for now was using Docker by commenting out the Rust part which was pretty slow. Even with these changes the Docker build is quite slow if I want to change some C code (since Docker caching does not work).

The Nix bits will likely be deprecated soon as they are not actively maintained. 

Building Omnigres as is,  without a container, is fairly easy: https://github.com/omnigres/omnigres?tab=readme-ov-file#building--using-extensions

With this process, testing changes takes few seconds. 

The Docker images are not meant for the development of Omnigres itself.

**Comment by UkuLoskit** [comment]
thanks! finally figured it out that I had create the extension after that executing the make target :)

**Comment by yrashk** [comment]
Indeed, this should be spelled out better. There are subtle reasons why it's not done by default (although we may reconsider them)

**Comment by UkuLoskit** [comment]
@yrashk been making some minor progress with this...
I created a table for certificates that would store the certificate and private key in PEM format, additional to the primary key (id) there would be a listener_id column.

I was attaching the OpenSSL context here to `accept_ctx.ssl_ctx`
https://github.com/omnigres/omnigres/blob/0e6be6587eb12ade5cac24428980b49cc5419e90/extensions/omni_httpd/http_worker.c#L239 and using this I was able to serve omni_httpd over SSL.

I'm wondering if this is the correct approach?
Also, it seems that the query results are read at later point than the query is executed. could they be read earlier? For attaching the SSL context I would need to read the query results at the point.

One option would be to add another join to the handlers query which would perform a left join with new certificates table on listener_id.

**Comment by yrashk** [comment]
@UkuLoskit, broadly seems right overall and with respect to the join.

As for reading the results at a later point, I don't think anything is stopping us from reading them twice. The results are effectively in an array anyway (https://github.com/omnigres/omnigres/blob/0e6be6587eb12ade5cac24428980b49cc5419e90/extensions/omni_httpd/http_worker.c#L265-L267), so perhaps we can read index-matched listener SSL key (if found) when we are setting up the listeners right there as well.

**Comment by rishi-jat** [comment]
@yrashk is this issue still open to work ?

**Comment by yuliuyi717-ux** [comment]
I can take this bounty and follow the requested pre-implementation plan review first.\n\nRough plan:\n1) add certificate configuration storage in  (table for cert/key material + listener association, with room for SNI matching extension),\n2) wire listener initialization to resolve cert config and initialize TLS context per listener,\n3) keep first iteration focused on static cert/key configuration (no ACME),\n4) add integration tests for HTTPS listener startup + basic request flow + misconfiguration cases.\n\nIf this direction matches maintainer expectations, I will proceed with an implementation PR in small commits.

**Comment by yuliuyi717-ux** [comment]
Correction: formatting issue in my previous comment.

Concrete plan proposal:
1) add certificate configuration storage in `omni_httpd` (table for cert/key material + listener association, with room for later SNI extension),
2) wire listener initialization to resolve cert config and initialize TLS context per listener,
3) keep first iteration focused on static cert/key configuration (no ACME),
4) add integration tests for HTTPS listener startup + basic request flow + misconfiguration cases.

If this plan aligns with maintainer expectations, I will proceed with implementation in small reviewable commits.

**Comment by yrashk** [comment]
We are no longer taking bounties
